Zend/zend_execute.c

```diff
@@ -1724,14 +1725,57 @@ ZEND_API ZEND_COLD void zend_wrong_string_offset_error(void)
 
 ZEND_API ZEND_COLD void ZEND_FASTCALL zend_deprecated_function(const zend_function *fbc)
 {
+	zend_string *message_suffix = zend_empty_string;
+
+	if (fbc->common.attributes != NULL) {
+		zend_attribute *deprecated = zend_get_attribute_str(fbc->common.attributes, "deprecated", sizeof("deprecated")-1);
+
+		if (deprecated->argc >= 1) {
+			zval message;
+
+			if (FAILURE != zend_get_attribute_value(&message, deprecated, 0, fbc->common.scope)) {
+				zval obj;
+
+				object_init_ex(&obj, zend_ce_deprecated);
+				zend_call_method_with_1_params(Z_OBJ_P(&obj), zend_ce_deprecated, &zend_ce_deprecated->constructor, "__construct", NULL, &message);
+				zval_ptr_dtor(&message);
+
+				if (EG(exception)) {
+					zval_ptr_dtor(&obj);
+					return;
+				}
+
+				zval *z;
+
+				if ((z = zend_read_property_ex(zend_ce_deprecated, Z_OBJ_P(&obj), ZSTR_KNOWN(ZEND_STR_MESSAGE), false, NULL)) == NULL) {
+					zval_ptr_dtor(&obj);
+					return;
+				}
+
+				message_suffix = zend_string_copy(Z_STR_P(z));
```

Comment:
Can we not simply inline the relevant part of `Deprecate::__construct()`? That should be much simpler and avoids an allocation.

Reply:
I find it useful to pass this through the constructor for consistent error handling of the parameter, but this might be refactored, when the compile time validation is what's done.

Reply:
I'd extract this to a common function then, rather than going through a useless object allocation.

Reply:
How would you go about this? Consistent error handling would also need to include a proper stack trace (see `Zend/tests/attributes/deprecated/009.phpt` and related tests) and going through ZPP.

This would effectively require me to do everything by hand, except possibly omitting the object allocation + pulling out the message from the property, no?

Reply:
I've added an optimization to bypass the object construction, if the input is a string. This should match the vast majority of cases and still provide proper error handling for the others.